### ISSUE TITLE
📎 Embedded language formatting

### ISSUE THREAD

**Issue by ah-yu** [biomejs/biome#3334]
## Preface

Some popular libraries allow code snippets in other languages to be embedded within JavaScript code. Users want to format these embedded code snippets within JavaScript to enhance the development experience.

## Design

Simply put, the idea is to extract the code snippets from template strings, format them using the respective language's formatter, and then replace them back into the template string.

### Handling Interpolation

We need to parse the entire template string and then format it based on the parsing results. However, template strings with interpolations are not valid CSS code (using CSS as an example here). Therefore, we need to preprocess the interpolations, turning the template string into a more valid CSS code. We plan to replace interpolations with a special string and then reinsert them after formatting.

To maximize parsing success, we chose to replace interpolations with grit metavariables. The reason for this choice you can find in https://github.com/biomejs/biome/pull/3228#issuecomment-2173404022

### Changes to the Public API

Since JavaScript formatters cannot directly format code in other languages, we need to use external tools to format these other languages' code. To achieve this, we designed a generic trait instead of relying on specific implementations, maximizing the decoupling between different language formatters.

```rust
enum JsForeignLanguage {
    Css,
}

trait JsForeignLanguageFormatter {
    fn format(&self, language: JsForeignLanguage, source: &str) -> FormatResult<Document>;
}
```

Then we can add a new parameter to the `format_node` function to pass in the formatter for other languages.

```diff
pub fn format_node(
    options: JsFormatOptions,
+   foreign_language_formatter: impl JsForeignLanguageFormatter,
    root: &JsSyntaxNode,
) -> FormatResult<Formatted<JsFormatContext>> {
    biome_formatter::format_node(
        root,
        JsFormatLanguage::new(options, foreign_language_formatter),
    )
}
```
## CLI

When formatting JavaScript files, we need to be aware of other languages' settings. For example, when formatting CSS code, we need to know the CSS formatter's settings.

## LSP

The LSP provides a feature called `format_range` that formats code snippets. This feature relies on [SourceMarker](https://github.com/biomejs/biome/blob/ce0068501c1c57e6382ca5e6d6b51fb45779eab5/crates/biome_formatter/src/lib.rs#L786-L791)s generated during the printing process. Generating a `SourceMarker` depends on the position information of tokens in the source code. This position information is contained in the following two `FormatElements`:

https://github.com/biomejs/biome/blob/ce0068501c1c57e6382ca5e6d6b51fb45779eab5/crates/biome_formatter/src/format_element.rs#L36-L50

Since the formatting of embedded languages is done by extracting, preprocessing, and then separately parsing and formatting them, the `source_position` in these two `FormatElement` is inaccurate, and the entire template string is handled as a whole. Therefore, I recommend erasing these inaccurate `source_position`. It is acceptable to erase them because the `format_range` function will still be able to find the `SourceMarker` closest to the range start and end. If there is a need to format parts of the embedded code in the future, we can revisit this issue.

## Tasks

- [ ] introduce grit metavariable in CSS
- [ ] change the public API
- [ ] preprocess template strings and handle the generated format elements.

**Comment by Sec-ant** [comment]
Great great work :rocket: . I'd like to add some additional notes. I will update them in this comment later.

**Comment by dyc3** [comment]
Something that I've realized while working on the html parser:

Transitioning into another language while parsing is not necessarily the hard part -- it's knowing when to transition back out instead of emitting an invalid syntax error.

Take this HTML for example, where javascript is embedded within `<script>`.

```html
<script>
const foo = "</script>";
</script>
```

I could see a more naive approach failing on the `<` operator in something simple like this:
```html
<script>
const foo = 5;
if (foo < 10) {
    console.log("foo");
}
</script>
```
It's not enough to simply look ahead for the end tokens for an embedded language. You can even see the syntax highlighting here on github fail for this case.

In short: We can only transition out of an embedded language when the embedded language is at the root level node.

**Comment by arendjr** [comment]
> ```
<script>
const foo = "</script>";
</script>
```

Does this actually work in any browser? I would expect it to actually be invalid syntax, so the correct result is for it to fail.

Basically, I think it’s the responsibility of the parent parser (the HTML one in this case) to determine where the snippet ends, and then you only need to hand the substring to the embedded parser. So finding the end token should never be the responsibility of the embedded parser.

**Comment by dyc3** [comment]
Actually, I just tested it in firefox and it didn't work.

There's also language in the spec that specifically talks about this too: https://html.spec.whatwg.org/#serialising-html-fragments:text

So that makes parsing way easier, we just scan until we hit `</script>`.

**Comment by glekner** [comment]
we really want to test out biome, this is one issue that is blocking us

**Comment by ematipico** [comment]
@glekner consider helping us, or upvote the issue. Just venting your frustration for something missing isn't fruitful, and it's usually considered almost spam or OT, because you just notify users subscribed to the issue.

**Comment by dyc3** [comment]
I've been tinkering with integrating embedded language parsing for HTML a little bit.

It's pretty clear at this point that we will need to mark certain areas in the CST as "this is where embedded languages can be". I created a `HtmlEmbeddedLanguage` node for this purpose. 

---

With my current understanding of how `biome_rowan` works, I believe it could be possible to kinda shove parent and child CSTs together in the same tree. This is how `SyntaxNode` is currently defined:
```rust
#[derive(Clone, PartialEq, Eq, Hash)]
pub struct SyntaxNode<L: Language> {
    raw: cursor::SyntaxNode,
    _p: PhantomData<L>,
}
```
The only thing that's technically enforcing the language constraint is the `PhantomData<L>`.

So I tried something like that:
```rust
for node in html.descendants() {
            if node.kind() == HtmlSyntaxKind::HTML_EMBEDDED_LANGUAGE && !node.text().is_empty() {
                let code = node.text().to_string();
                dbg!("Found embedded language", &code);

                let js_root = biome_js_parser::parse(
                    &code,
                    // TODO: determine the correct options
                    JsFileSource::js_script(),
                    JsParserOptions::default(),
                );
                let as_html = HtmlSyntaxNode::from_foriegn(js_root.syntax());
                // TODO: then replace this node with as_html
            }
        }
```
It works (sorta), but it panics when you try to print it (because the SyntaxKind is different).
Of course there are several things wrong with this approach. But it gave me a couple of other ideas for implementation.

One option is that we could have a "SuperSyntaxNode" that could be any language. For formatting, when we reach a node for that language, we just use the formatter for that language to format the node. (Obviously not that simple, I'm handwaving the details for now.)

Another option could be that we have a special "transition node" with a marker to indicate what language to transition into. This node would only ever have a single child, which would be the root node for the embedded language. I think this makes the most sense, but I have absolutely no clue how we would represent that in rowan.

**Comment by arendjr** [comment]
Might be good to know that we also have `GritTargetSyntaxNode` which is also intended to be agnostic over language type. So it’s actually an enum where the variants are language-specific syntax kinds (JS is the only variant for now). We could it move it from there and rename it if it’s useful for other things too.

**Comment by arendjr** [comment]
That said, it feels like a “transition node” might actually be the better choice here. I guess you may not need to perform any special Rowan tricks if you just put the embedded tree in the transition node’s data?

**Comment by dyc3** [comment]
@arendjr I didn't find a `GritTargetSyntaxNode`, did you mean `GritTargetLanguageNode`?

**Comment by arendjr** [comment]
Ah, yes indeed!

**Comment by dyc3** [comment]
I've hit another blocker. Our parsing infra expects all `SyntaxKind` to map to a specific `u16` value.

```rust
impl From<u16> for HtmlSyntaxKind {
    fn from(d: u16) -> HtmlSyntaxKind {
        assert!(d <= (HtmlSyntaxKind::__LAST as u16));
        let k = unsafe { std::mem::transmute::<u16, HtmlSyntaxKind>(dbg!(d)) };
    }
}
```

This assumption causes lots of problems when trying to force trees from 2 different languages in the same tree. It doesn't seem like its easy to add a new `SyntaxKind::FOREIGN_TOKEN(u16)` to handle these in the same tree.

I'm starting to think that the better approach would be to just do the embedded language parsing after the initial parsing pass and just keep them as separate trees.

**Comment by arendjr** [comment]
Right, I think that's kinda what I was getting at with the "put the embedded tree in the transition node’s data". The way I would expect this to work:

1. Parent (HTML) parser discovers a "foreign node" that contains embedded content.
2. Parent parser continues looking for the closing tag. Now it knows the range of the embedded content.
3. Parent parser invokes the correct child (CSS, JS) parser.
4. Child parser produces a separate tree.
5. Parent parser stores the produced tree in the data of the foreign node.

**Comment by ematipico** [comment]
Yes, we can't do multiple parsing in the same phase because of `L: Language`, which is one of the greatest constraints we have in the code base (good or bad).

Theoretically, what @arendjr suggested could be done, but this would involve mutating an existing CST using the same mutations we use for the analyzer, which could be slow, but maybe we could overlook it.

I'll have to study the problem a bit more when I have more time.

**Comment by orimay** [comment]
We also need to be able to indent script and style contents (prettier's property `vueIndentScriptAndStyle` allows that):
```vue
<script>
  const foo = 5;
  if (foo < 10) {
    console.log("foo");
  }
</script>
```

**Comment by loicnestler** [comment]
Any updates on this? Is there an ETA for embedded language formatting?

**Comment by rory-orennia** [comment]
Just wanted to add something that currently isn't handled by the partial svelte support: multiple script tags.  You might do something like this in a file to create a type that can be referenced by other code outside of the component:
```svelte
<script context="module" lang="ts">
export interface ColumnFiltered {
  value: string;
  label: string;
}
</script>

<script lang="ts">
  export let columnsFiltered: ColumnFiltered[] | null;
</script>
```
As you can see the first script tag is formatted, but the 2nd is not. (look at the indenting).  I'm hoping that's an easy fix where you iterate over all script tags instead of just the first one you see.  But if it's not then I wanted to make sure this case is marked down somewhere in the Svelte support roadmap.  Thanks!

**Comment by siketyan** [comment]
@dennemark Yes, we don't support any type of embed languages yet, except the frontmatter part of Vue, Svelte, or Astro.

**Comment by conartist6** [comment]
I just wanted to mentioned something from my experience with this kind of thing: there are actually two subtly different kinds of embedding, which I refer to as "embedded in" and "embedded on".

Regex, as a language, is embedded in JS. You can tell that the regex language is in JS because JS is aware of regex syntax. The example that shows this is `/[/]/`, where the inner `/` would terminate parsing if the JS parser wasn't aware of the meaning of `[` in regex.

Markdown code blocks are an example of embedding a lanugage *on* another language. A markdown code block is basically just a big string literal. You parse it, then you can optionally take a second pass to apply parsing rules to the content of the string, and maybe attach that parse result to the string.

**Comment by eveeifyeve** [comment]
@ah-yu are you continually wanting to work on this? if not please un selfassign yourself so we can at least know who is working on it.

**Comment by ematipico** [comment]
He hasn't been active for a while, so it's safe to say that he hasn't had time to work on it. Unassigned.

**Comment by mantou132** [comment]
I would like to ask, will all embedded languages ​​that can be highlighted in the future be formatted?

e.g: plugin highlight https://github.com/mantou132/gem/blob/main/crates/zed-plugin-gem/src/injections.scm

**Comment by arendjr** [comment]
@mantou132 There's no direct relation between highlighting and formatting. The only languages we can format are languages supported by Biome, so probably no. We are open to extending support to more of such languages though!

**Comment by eveeifyeve** [comment]
> [@mantou132](https://github.com/mantou132) There's no direct relation between highlighting and formatting. The only languages we can format are languages supported by Biome, so probably no. We are open to extending support to more of such languages though!

This could be done through the plugin system.

**Comment by JongSikLim** [comment]
Do you have a plan for this issue? I'm just waiting for migration from ESLint + prettier. This is a blocker for me.

**Comment by ematipico** [comment]
We appreciate the interest in the issue, and I would appreciate some more understanding regarding how things work for us (and generally open source):
- It is planned https://biomejs.dev/blog/roadmap-2025/#-2025-roadmap
- The roadmap is mostly a "what we would like to ship", however it doesn't mean we will be able to
- the majority of people working on the project do that in their spare time, and they are also free to work on the part that like
- we understand something is a blocker,  so please help to ship it faster or consider a sponsorship so we can employ people to work on that. If those aren't opinions available to you, please have patience

**Comment by hbmartin** [comment]
This most important use-case for me is working with Svelte. Could Svelte possibly be its own domain?

**Comment by conartist6** [comment]
A follow-up to my earlier comment about embedded-on and embedded-in. In the current tree structure regex literals in Javascript files are stored as if this is an embedded-on relationship, i.e. the regex pattern has no parse tree and presented as a string, as is normal and correct for languages embedded on other languages.

Regex, as I mentioned previously, is embedded *in* Javascript though, in other words it's a completely inseparable part of the syntax and semantics of Javascript, and the parser should treat it as such, including parsing its nodes and presenting a regex as a normal part of the syntax tree with its own internal node types.

Most likely if you were to start doing this work you would quickly find that you have name locality issues. The symptom of this will be all the names for node types beginning with `Regex` e.g. you might end up with `RegexPattern` and `RegexCharacterClass` and `RegexCharacter` and `RegexFlags` etc. I think of this as a code smell basically, because we don't normally have to prefix the name of the language onto the name of every node -- instead the language is the context for the nodes.

That's why I think this issue has a chance to create the possibility of fixing the nature of regex embedding: once you can transition between different nested linguistic contexts in your data structure before you can have these nice contextually appropriate node names like `Character` and `Flags`.

If I could recommend a particular architectural patterning to solve this I'd recommend creating some kind of binding primitive -- a linking object which is neither strictly part of the parent nor part of the child. The reasons for doing this are the same reasons that connections are important in GraphQL. As it was explained to me when I was at Facebook, in a graph like a social graph some information you need to store is not about any one person, but rather about how two or more people relate to each other. Where in a graph of people would you store the date when two individuals became friends with each other, for example? It's not information directly about either person. If you were to store the data on the data model for each person you'd have to duplicate the same data -- once on each side of the friendship, and in each person the data would have to be a mapping from `friend -> date`. Instead the approach Facebook chose was essentially to make "friendship" its own object in the data model on which they could store all the kinds of metadata about the relationship to users had to each other.

The way you spot these kinds of distinctions is by considering which attributes are intrinsic (essential to the identity of a thing) and which are extrinsic (products of circumstance). My argument is that a node's intrinsic language is immutable over time: it's just the language that parsed that node, and therefore the best way to introduce information into the parse tree about language embedding is to put it not in a child node but between the parent and child.

As one last clarification, an example of design I don't like is XML: `<Language.Node></Language.Node>`. It implies that `Language.` is essential to what a node is, which is untrue: it's actually extrinsic, which is to say that part of the part of the definition of this node will change if it is in a different parent. An extrinsic presentation for XML would be more like: `(Language: <Node></Node>)`

**Comment by difosfor** [comment]
In the mean time could the documentation of Biome be improved? I feel like currently it implies to be almost completely compatible with Prettier, i.e:

> scores [97% compatibility with Prettier](https://console.algora.io/challenges/prettier)

Only when you click through on that can you read in the small print that embedded language support is outside of that scope. Similarly Markdown support etc are missing when compared to Prettier.

Nowhere in the documentation can I find a clear list of missing features for people that are migrating:
- [Migrate from ESLint and Prettier](https://biomejs.dev/guides/migrate-eslint-prettier/)
- [Differences with Prettier](https://biomejs.dev/formatter/differences-with-prettier/)
- [Language support](https://biomejs.dev/internals/language-support/)

-- 

TL;DR: I suggest on the front page after the 97% compatibility with Prettier link you mention a brief list of unsupported feature names between parentheses or so. And you can link that to the Differences with Prettier page where you could detail those missing features and perhaps link to the related issues here on GitHub and/or how to contribute.

**Comment by Zearin** [comment]
(After looking at the other Issues that exist, I think this is the most appropriate place for me to say the following. Apologies in advance if I missed a more appropriate place to voice this.)

----

I just noticed the following in an HTML file I’m working on in VSCodium:

<img width="757" height="143" alt="Image" src="https://github.com/user-attachments/assets/fb3d6b68-e72c-4841-8cb2-feeaf7a35765" />

The thing is… the statement that frontmatter is “only valid inside Astro files” is false.

Back in the early 2010s, I was using the excellent [Metalsmith](https://github.com/metalsmith/metalsmith) static site generator to build websites, and it heavily featured frontmatter as a flexible tool for a “roll your own” config/workflow/whatever.

These days, I use [Eleventy](https://github.com/11ty/eleventy), but it also features frontmatter as an incredibly flexible, useful tool for developers who like the freedom that frontmatter provides.

**Please don’t treat frontmatter as an Astro-specific feature.**  Ideally, offer opt-in support for it at a general- or language-specific level.  (Less-than-ideally, at least update the parser’s message to say something like “Biome currently only supports frontmatter in Astro files”.)

**Comment by dyc3** [comment]
@Zearin please file a new issue and explain your use case

**Comment by lucsoft** [comment]
i don't quite understand why this was closed? by mistake? Or are we not getting Embedded language formatting?

**Comment by ematipico** [comment]
> i don't quite understand why this was closed? by mistake? Or are we not getting Embedded language formatting?

Because it's been implemented since v2.3

We have the infrastructure and we started to flash it out via:
- html experimental full support
- experimental JavaScript snippets (graphql and CSS in js)

So there's nothing to track. If you experience bugs with features I listed, file a bug

**Comment by karlhorky** [comment]
@ematipico thanks, amazing! 

Is there anything tracking SQL-in-JS (eg. with [Postgres.js](https://github.com/porsager/postgres))? Or should an issue be made for this too?

**Comment by ematipico** [comment]
Biome doesn't support SQL, and it isn't part of our roadmap for the time being.

You can open a discussion to gather upvotes

**Comment by blacelle** [comment]
I read from https://www.biomejs.cn/en/internals/language-support/:

> Since version v2.3.0, Biome supports Vue, Svelte and Astro file out of the box. This means that Biome is able to format and lint the HTML, CSS and JavaScript parts of the files.

Does it mean support for .js files with HTML variables are not supported? With prettier, one could flag/hint some variable/value as html, with with `/* HTML */`. See https://github.com/prettier/prettier/blob/2247ce1aab604bf1b95b3d010c08d48b003286c7/src/language-js/embed.js#L548-L565

Hence managing a .js file like:

```
export default {
	template: /* HTML */ `
		<div>hello</div>
	`,
};
```
(In practise, I rely on [VueJS from CDN](https://vuejs.org/guide/quick-start.html#using-vue-from-cdn), hence through .js files ; but the `/* HTML */` syntax covers additional cases).

- Is it already supported? (I guess not as I can not get it working)
- Is it planned? (I did not find anything around this)
- Should I open a dedicated ticket?

Thanks
